File: sdk/typescript/src/scanners/AhoCorasick.ts
// ===========================================================================
// OPT-003: Aho-Corasick Multi-Pattern String Matching
// Pure TypeScript implementation. No external dependencies.
// Single-pass O(n + m) text scanning for all registered patterns.
// ===========================================================================

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface AhoCorasickMatch {
  /** Index of the matched pattern in the original patterns array */
  patternIndex: number;
  /** Start position in the searched text */
  start: number;
  /** End position in the searched text (exclusive) */
  end: number;
  /** The matched text */
  text: string;
}

// ---------------------------------------------------------------------------
// Trie Node
// ---------------------------------------------------------------------------

interface TrieNode {
  /** Children keyed by character */
  children: Map<string, TrieNode>;
  /** Failure link (suffix pointer) */
  fail: TrieNode | null;
  /** Pattern indices that end at this node (including via dictionary suffix links) */
  output: number[];
  /** Depth in the trie (= length of the string represented by this node) */
  depth: number;
}

function createNode(depth: number): TrieNode {
  return {
    children: new Map(),
    fail: null,
    output: [],
    depth,
  };
}

// ---------------------------------------------------------------------------
// Aho-Corasick Automaton
// ---------------------------------------------------------------------------

export class AhoCorasick {
  private readonly root: TrieNode;
  private readonly patterns: string[];
  private readonly caseSensitive: boolean;
  private built = false;

  /**
   * @param patterns Array of literal strings to search for.
   * @param caseSensitive If false (default), matching is case-insensitive.
   */
  constructor(patterns: string[], caseSensitive = false) {
    this.patterns = patterns;
    this.caseSensitive = caseSensitive;
    this.root = createNode(0);
    this.root.fail = this.root;

    this.buildTrie();
    this.buildFailureLinks();
    this.built = true;
  }

  /**
   * Search text in a single pass. Returns all matches with pattern index,
   * start position, end position, and matched text.
   */
  search(text: string): AhoCorasickMatch[] {
    if (!this.built || this.patterns.length === 0) return [];

    const matches: AhoCorasickMatch[] = [];
    let current = this.root;
    const searchText = this.caseSensitive ? text : text.toLowerCase();

    for (let i = 0; i < searchText.length; i++) {
      const ch = searchText[i];

      // Follow failure links until we find a transition or reach root
      while (current !== this.root && !current.children.has(ch)) {
        current = current.fail!;
      }

      if (current.children.has(ch)) {
        current = current.children.get(ch)!;
      }
      // else: current remains root

      // Collect all outputs at this node (including via dictionary suffix links)
      let outputNode: TrieNode | null = current;
      while (outputNode !== null && outputNode !== this.root) {
        for (const patIdx of outputNode.output) {
          const patLen = this.patterns[patIdx].length;
          const start = i - patLen + 1;
          matches.push({
            patternIndex: patIdx,
            start,
            end: i + 1,
            text: text.substring(start, i + 1),
          });
        }
        outputNode = outputNode.fail!;
        // Prevent infinite loop at root
        if (outputNode === this.root) break;
      }
    }

    return matches;
  }

  /**
   * Return the number of patterns registered.
   */
  get patternCount(): number {
    return this.patterns.length;
  }

  // -----------------------------------------------------------------------
  // Build Phase
  // -----------------------------------------------------------------------

  private buildTrie(): void {
    for (let i = 0; i < this.patterns.length; i++) {
      const pattern = this.caseSensitive ? this.patterns[i] : this.patterns[i].toLowerCase();
      let current = this.root;

      for (let j = 0; j < pattern.length; j++) {
        const ch = pattern[j];
        if (!current.children.has(ch)) {
          current.children.set(ch, createNode(j + 1));
        }
        current = current.children.get(ch)!;
      }

      // Mark end of pattern
      current.output.push(i);
    }
  }

  private buildFailureLinks(): void {
    // BFS to compute failure links
    const queue: TrieNode[] = [];

    // Depth-1 nodes: failure link → root
    for (const child of this.root.children.values()) {
      child.fail = this.root;
      queue.push(child);
    }

    while (queue.length > 0) {
      const current = queue.shift()!;

      for (const [ch, child] of current.children) {
        queue.push(child);

        // Compute failure link for child
        let failNode = current.fail!;
        while (failNode !== this.root && !failNode.children.has(ch)) {
          failNode = failNode.fail!;
        }

        child.fail = failNode.children.has(ch) ? failNode.children.get(ch)! : this.root;

        // Merge dictionary suffix link outputs
        if (child.fail !== child) {
          child.output = child.output.concat(child.fail.output);
        }
      }
    }
  }
}

File: sdk/python/dynaep/scanners/aho_corasick.py
"""
OPT-003: Aho-Corasick Multi-Pattern String Matching (Python)
Pure Python implementation. No external dependencies.
Falls back to pyahocorasick if available for better performance.
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field
from typing import Optional

try:
    import ahocorasick as _pyac
    _HAS_PYAHOCORASICK = True
except ImportError:
    _HAS_PYAHOCORASICK = False


@dataclass
class AhoCorasickMatch:
    pattern_index: int
    start: int
    end: int
    text: str


class _TrieNode:
    __slots__ = ("children", "fail", "output", "depth")

    def __init__(self, depth: int = 0):
        self.children: dict[str, _TrieNode] = {}
        self.fail: Optional[_TrieNode] = None
        self.output: list[int] = []
        self.depth: int = depth


class AhoCorasick:
    """
    Pure Python Aho-Corasick automaton.
    Single-pass O(n + m) text scanning for all registered patterns.
    """

    def __init__(self, patterns: list[str], case_sensitive: bool = False):
        self._patterns = patterns
        self._case_sensitive = case_sensitive
        self._root = _TrieNode(0)
        self._root.fail = self._root
        self._build()

    def search(self, text: str) -> list[AhoCorasickMatch]:
        """Search text in a single pass. Returns all matches."""
        if not self._patterns:
            return []

        matches: list[AhoCorasickMatch] = []
        current = self._root
        search_text = text if self._case_sensitive else text.lower()

        for i, ch in enumerate(search_text):
            while current is not self._root and ch not in current.children:
                current = current.fail  # type: ignore

            if ch in current.children:
                current = current.children[ch]

            output_node: Optional[_TrieNode] = current
            while output_node is not None and output_node is not self._root:
                for pat_idx in output_node.output:
                    pat_len = len(self._patterns[pat_idx])
                    start = i - pat_len + 1
                    matches.append(AhoCorasickMatch(
                        pattern_index=pat_idx,
                        start=start,
                        end=i + 1,
                        text=text[start:i + 1],
                    ))
                output_node = output_node.fail
                if output_node is self._root:
                    break

        return matches

    @property
    def pattern_count(self) -> int:
        return len(self._patterns)

    def _build(self) -> None:
        self._build_trie()
        self._build_failure_links()

    def _build_trie(self) -> None:
        for i, pattern in enumerate(self._patterns):
            p = pattern if self._case_sensitive else pattern.lower()
            current = self._root
            for j, ch in enumerate(p):
                if ch not in current.children:
                    current.children[ch] = _TrieNode(j + 1)
                current = current.children[ch]
            current.output.append(i)

    def _build_failure_links(self) -> None:
        queue: deque[_TrieNode] = deque()

        for child in self._root.children.values():
            child.fail = self._root
            queue.append(child)

        while queue:
            current = queue.popleft()

            for ch, child in current.children.items():
                queue.append(child)

                fail_node = current.fail
                while fail_node is not self._root and ch not in fail_node.children:
                    fail_node = fail_node.fail  # type: ignore

                child.fail = fail_node.children.get(ch, self._root)
                if child.fail is child:
                    child.fail = self._root

                child.output = child.output + child.fail.output
